### PR TITLE
Task scheduler with a background thread.

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/task_scheduler.h [hunk 1]
```diff
@@ -33,7 +33,12 @@ struct aws_task {
     struct aws_linked_list_node node;
     struct aws_priority_queue_node priority_queue_node;
     const char *type_tag;
-    size_t reserved;
+
+    /* honor the ABI compat */
+    union {
+        bool scheduled;
```
Comment (justinboswell) [inline include/aws/common/task_scheduler.h:39]: Do you want to bitfield this?

File: include/aws/common/task_scheduler.h [hunk 2]
```diff
@@ -33,7 +33,12 @@ struct aws_task {
...
+        size_t reserved;
+    } abi_extension;
```
Comment (justinboswell) [inline include/aws/common/task_scheduler.h:41]: This union could be anonymous and still achieve ABI compat.

**Review by justinboswell — COMMENTED** [review]
So this will give us 1 background thread, which is useful, but I suspect in the use case we're looking at it's going to back up. Are we going to end up with a thread pool at some point, holding a collection of these?

**Comment by justinboswell** [comment]
Also, I am totally going to replace aws_message_bus's async guts with this.

File: include/aws/common/task_scheduler.h (hunk 1 above)
Comment (JonathanHenson) [reply include/aws/common/task_scheduler.h:39]: We can, but we don’t have to til we start using the other 3 bytes.

File: include/aws/common/task_scheduler.h (hunk 2 above)
Comment (JonathanHenson) [reply include/aws/common/task_scheduler.h:41]: Yeah but it won’t compile because that’s a c12 feature apparently.

**Comment by JonathanHenson** [comment]
> So this will give us 1 background thread, which is useful, but I suspect in the use case we're looking at it's going to back up. Are we going to end up with a thread pool at some point, holding a collection of these?

That’s the idea.... but that part isn’t hard, and I’m assuming you’re already doing that. Regardless it’s a separate implementation from this.

**Review by justinboswell — APPROVED** [review]

**Pull request merged**